spec.html

```diff
@@ -8631,18 +8631,58 @@ <h1>WithBaseObject ( )</h1>
       <emu-clause id="sec-object-environment-records">
         <h1>Object Environment Records</h1>
         <p>Each <dfn>object Environment Record</dfn> is associated with an object called its <em>binding object</em>. An object Environment Record binds the set of string identifier names that directly correspond to the property names of its binding object. Property keys that are not strings in the form of an |IdentifierName| are not included in the set of bound identifiers. Both own and inherited properties are included in the set regardless of the setting of their [[Enumerable]] attribute. Because properties can be dynamically added and deleted from objects, the set of identifiers bound by an object Environment Record may potentially change as a side-effect of any operation that adds or deletes properties. Any bindings that are created as a result of such a side-effect are considered to be a mutable binding even if the Writable attribute of the corresponding property has the value *false*. Immutable bindings do not exist for object Environment Records.</p>
-        <p>Object Environment Records created for `with` statements (<emu-xref href="#sec-with-statement"></emu-xref>) can provide their binding object as an implicit *this* value for use in function calls. The capability is controlled by a _withEnvironment_ Boolean value that is associated with each object Environment Record. By default, the value of _withEnvironment_ is *false* for any object Environment Record.</p>
+        <p>Object Environment Records created for `with` statements (<emu-xref href="#sec-with-statement"></emu-xref>) can provide their binding object as an implicit *this* value for use in function calls. The capability is controlled by a Boolean [[WithEnvironmentFlag]] field.</p>
+        <p>Object Environment Records have the additional state fields listed in <emu-xref href="#table-additional-fields-of-object-environment-records"></emu-xref>.</p>
+        <emu-table id="table-additional-fields-of-object-environment-records" caption="Additional Fields of Object Environment Records">
+          <table>
+            <tbody>
+            <tr>
+              <th>
+                Field Name
+              </th>
+              <th>
+                Value
+              </th>
+              <th>
+                Meaning
+              </th>
+            </tr>
+            <tr>
+              <td>
+                [[BindingObject]]
+              </td>
+              <td>
+                Object
+              </td>
+              <td>
+                The binding object of this Environment Record.
+              </td>
+            </tr>
+            <tr>
+              <td>
+                [[WithEnvironmentFlag]]
```

Comment:
Nit: stylistically I'd prefer `IsWithEnvironment`.

Reply:
"HasEnvironment"? "is with" reads weird.

Reply:
Here "With" is an adjective, not a preposition: that is, this should be read as "is an environment for a WithStatement". "HasEnvironment" would mean something else entirely.

Reply:
I initially thought of `IsWithEnvironment`, but decided that `WithEnvironmentFlag` read better. Mind you, they're both sub-optimal in that there's no such thing as a "With Environment". Some other possibilities:
- `IsForAWithStatement`
- `ProvidesAWithBaseObject`

Reply:
The spec doesn't have a formal notion of a "With Environment", but readers are likely to understand it, I think. I still like `IsWithEnvironment` best of those options, but let's see if @michaelficarra or @syg have opinions.

Reply:
(Conceivably, we could define a With Environment Record as a 'sub-type' of object Environment Record, but I don't think it'd be worth the bother.)

Reply:
Naming it `[[UseSymbolUnscopables]]` should also work.

Reply:
Naming it `[[UseSymbolUnscopables]]` should also work.

Reply:
`[[UseSymbolUnscopables]]` isn't quite right because this boolean also affects `WithBaseObject`, which is used in `EvaluateCall` (so that if you do `let obj = { o(){'use strict'; console.log(this); } }; with (obj) o();` it prints `obj` but if you do `globalThis.o = function(){'use strict'; console.log(this); }; o();` it prints `undefined`).

Editors discussed this on the editor call and are in favor of `IsWithEnvironment`. It doesn't read all that well, but it's more consistent with other slot names and it's a sufficiently obscure slot that the awkwardness will not be exposed to many people.

Reply:
Done.